### PR TITLE
MODDATAIMP-352: Implement source reader for EDIFACT files.

### PR DESCRIPTION
**Pull request by Aliaksandr-Fedasiuk**
## Purpose
This PR is 1st implementation of EDIFACT reader for separating original EDIFACT files to chunks. 

## Approach
This was accomplished by adding the EdifactReader implementation of the SourceReader interface. 

## Learning
More info: [MODDATAIMP-352](https://issues.folio.org/browse/MODDATAIMP-352)

### CODE REVIEW

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      InputStream stream = new FileInputStream(file);
+      reader = factory.createEDIStreamReader(stream);
+      if (reader.next() == EDIStreamEvent.START_INTERCHANGE) {
+        delimiters = reader.getDelimiters();
+        edifactParser = new EdifactParser(delimiters);
+      }
+    } catch (IOException | EDIStreamException e) {
+      String errorMessage = "Can not initialize reader. Cause: " + e.getMessage();
+      LOGGER.error(errorMessage);
+      throw new IllegalArgumentException(errorMessage);
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/service/processing/reader/EdifactReader.java:50]: we have a custom Reader exception that is used as a wrapper over the IOException, I think it is applicable here as well

File: src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  public void handle(String data) {
+    String content = parser.getHeader()
+      + parser.getInvoiceBody() + data + getSegmentSeparator();
+    parser.getInitialRecords().add(getInitialRecord(content));
+    parser.cleanInvoiceBody();
+  }
+
+  @Override
+  public String getContent() {
+    throw new IllegalArgumentException("Not implemented.");
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java:24]: UnsupportedOperationException?

File: src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  @Override
+  public void cleanContent() {
+    throw new IllegalArgumentException("Not implemented.");
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java:29]: same here

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  @Override
+  public List<InitialRecord> next() {
+    try {
+      Files.lines(file.toPath(), DEFAULT_CHARSET).map(l -> l.split(edifactParser.getSegmentSeparator()))
+        .flatMap(Arrays::stream).forEach(edifactParser::handle);
+    } catch (IOException e) {
+      String errorMessage = "Error during handling the file: " + file.getName();
+      LOGGER.error(errorMessage, e);
+      throw new IllegalArgumentException(errorMessage);
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/service/processing/reader/EdifactReader.java:62]: same here

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java (hunk 1 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/EdifactReader.java]: Fixed.

File: src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java (hunk 2 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java]: Fixed.

File: src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java (hunk 3 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/model/EdifactInvoiceFinishState.java]: Fixed.

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java (hunk 4 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/EdifactReader.java]: Fixed.

**Review by KaterynaSenchenko — APPROVED** [review]

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  private EdifactParser edifactParser;
+  private File file;
+  private Map<String, Character> delimiters = null;
+
+  public EdifactReader(File file) {
+    try {
+      this.file = file;
+      EDIInputFactory factory = EDIInputFactory.newFactory();
+      InputStream stream = new FileInputStream(file);
+      reader = factory.createEDIStreamReader(stream);
```
Comment (RuslanLavrov) [inline src/main/java/org/folio/service/processing/reader/EdifactReader.java:42]: hello @Aliaksandr-Fedasiuk, put please the `reader` creation into a try-with-resources statement to make sure that the  opened file stream is closed

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    } catch (EDIStreamException e) {
+      LOGGER.error("Can not initialize reader. Cause: " + e.getMessage());
+      throw new RecordsReaderException(e);
+    }
+  }
+
+  @Override
+  public List<InitialRecord> next() {
+    try {
+      Files.lines(file.toPath(), DEFAULT_CHARSET).map(l -> l.split(edifactParser.getSegmentSeparator()))
```
Comment (RuslanLavrov) [inline src/main/java/org/folio/service/processing/reader/EdifactReader.java:59]: could you put pls. the `Files#lines()` method into a try-with-resources as well

**Review by RuslanLavrov — APPROVED** [review]

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java (hunk 5 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/EdifactReader.java:42]: Fixed.

File: src/main/java/org/folio/service/processing/reader/EdifactReader.java (hunk 6 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/service/processing/reader/EdifactReader.java]: Fixed.

**Review by VRohach — APPROVED** [review]

**Review by OleksiiKuzminov — APPROVED** [review]

**Pull request merged**